docs/02-app/01-building-your-application/02-data-fetching/02-server-actions-and-mutations.mdx

```diff
@@ -0,0 +1,831 @@
+    </form>
+  )
+}
+```
+
+```jsx filename="app/ui/signup.js" switcher
+'use client'
+
+import { useFormState } from 'react-dom'
+import { createUser } from '@/app/actions'
+
+const initialState = {
+  message: null,
+}
+
+export function Signup() {
+  const [state, formAction] = useFormState(createUser, initialState)
+
+  return (
+    <form action={formAction}>
+      <label htmlFor="email">Email</label>
+      <input type="text" id="email" name="email" required />
+      {/* ... */}
+      <p aria-live="polite" className="sr-only">
+        {state?.message}
+      </p>
+      <button>Sign up</button>
+    </form>
+  )
+}
+```
+
+#### Optimistic Updates
+
+You can use the React [`useOptimistic`](https://react.dev/reference/react/useOptimistic) hook to optimistically update the UI before the Server Action finishes, rather than waiting for the response:
+
+```tsx filename="app/page.tsx" switcher
+'use client'
+
+import { useOptimistic } from 'react'
+import { send } from './actions'
+
+type Message = {
+  message: string
+}
+
+export function Thread({ messages }: { messages: Message[] }) {
+  const [optimisticMessages, addOptimisticMessage] = useOptimistic<Message[]>(
+    messages,
+    (state: Message[], newMessage: string) => [
+      ...state,
+      { message: newMessage },
+    ]
+  )
+
+  return (
+    <div>
+      {optimisticMessages.map((m, k) => (
+        <div key={k}>{m.message}</div>
+      ))}
+      <form
+        action={async (formData: FormData) => {
+          const message = formData.get('message')
+          addOptimisticMessage(message)
+          await send(message)
+        }}
+      >
+        <input type="text" name="message" />
+        <button type="submit">Send</button>
+      </form>
+    </div>
+  )
+}
+```
+
+```jsx filename="app/page.jsx" switcher
+'use client'
+
+import { useOptimistic } from 'react'
+import { send } from './actions'
+
+export function Thread({ messages }) {
+  const [optimisticMessages, addOptimisticMessage] = useOptimistic(
+    messages,
+    (state, newMessage) => [...state, { message: newMessage }]
+  )
+
+  return (
+    <div>
+      {optimisticMessages.map((m) => (
+        <div>{m.message}</div>
+      ))}
+      <form
+        action={async (formData) => {
+          const message = formData.get('message')
+          addOptimisticMessage(message)
+          await send(message)
+        }}
+      >
+        <input type="text" name="message" />
+        <button type="submit">Send</button>
+      </form>
+    </div>
+  )
+}
+```
+
+#### `<button>` and `<input>` Elements
+
+You can invoke a Server Action from other form elements like `<button>`, `<input type="submit">`, and `<input type="image">`. These elements accept the `formAction` prop. For example, you can create a specific `<button>` element for saving a post draft in addition to publishing it. See the [React `<form>` docs](https://react.dev/docs/forms#handling-multiple-submit-buttons) for examples.
+
+### Non-form Elements
```

Comment:
@shuding would appreciate your thoughts on the examples for **Event Handlers** and **`useEffect`**. I tried to find examples that were simple enough for people to understand. But unsure if it's the correct usage.

Reply:
One example I have is about things that are not submitted via buttons. For example a checkbox that mutates the data when toggling:

```js
<input type="checkbox" onChange={async (e) => {
  await action(e.target.value)
}}/>
```

That's useful for optimistic forms. Same for `type="select"`, `type="radio"` and `type="input"` (e.g. auto-saving content with input debouncing) or even `<textarea>`s.

Another general use case with event handlers is for use cases that are usually not related to "forms", like triggering actions on `hover` events for advertising purposes. I don't have specific example here though but the ability is necessary.

For `useEffect`, the action might be depending on some global events:
- `onKeyDown` if you are adding keyboard shortcuts to your app. Imaging pressing `L` to like a tweet.
- `onScroll` for infinite loading related UX.

Or just triggered automatically, e.g. when this component is rendered (or the user has stayed for more than 5 minutes), increase the view counter in the DB.

Reply:
Thank you @shuding 🙏🏼 This is very helpful!